### PR TITLE
feat(android/app): Add menu to adjust keyboard height

### PR DESCRIPTION
**Pull request by darcywong00**
Addresses #3225 for Android

This adds a new "Settings" menu to adjust the keyboard height. The new menu allows the user to adjust the keyboard height for the current orientation (portrait vs landscape). These adjusted heights are saved as a SharedPreference.
The "RESET TO DEFAULTS" button clears out the saved heights so the default layout dimensions are used.

* Also adds a new API call ~`KMManager.setKeyboardHeight()`~ `KMManager.applyKeyboardHeight()` so the adjusted height gets passed to KeymanWeb (`setOskHeight`). The Javascript call already exists in keyboard.html, but they currently were called from `KMKeyboard.onResume()` and `KMKeyboard.onConfigurationChanged()`

TODO for separate PR - Use a static KeymanWeb OSK image for the new menu

### Screenshots

"Settings" menu
settings

"Adjust Keyboard Height" with the default keyboard height
adjust default

"Adjust Keyboard Height" with a bigger keyboard height
adjust

"Updated keyboard OSK"
adjusted osk

## User Testing
@keymanapp/testers 

### Setup
* Load the test build
* From the "Get Started" menu, enable Keyman as a system-wide keyboard and set Keyman as a default system keyboard.
* Note on the menu for adjusting keyboard height, given a default height (from the layout dimens.xml files, the range of values the keyboard height can be adjusted is:
`default height / 2` <= `adjusted height` <= `2 * default height`

### Adjust Keyboard Height (portrait orientation)
* Have the device in portrait orientation
* Go to "Settings" --> "Adjust keyboard height"
* Note the current keyboard height (default value)
* On the menu, drag the keyboard to either increase or decrease the keyboard height, and then release your finger
* Verify the red keyboard height adjusts accordingly
* Hit the back button until you're back in the Keyman app text area
* Verify the in-app keyboard height is adjusted per the menu
* On the device, open a separate app (like Chrome)
* Select an input field and display the Keyman keyboard
* Verify the system keyboard height is also adjusted per the menu
* In the Keyman app, go back to the "Adjust keyboard height" menu
* Press the "RESET TO DEFAULTS" button
* Verify the red keyboard height adjusts back to the default height
* Hit the back button until you're back in the Keyman app text area
* Verify the in-app keyboard height is adjusted per the menu

### Adjust Keyboard Height (landscape orientation)
Repeat the steps from the portrait orientation test, but with the device in landscape orientation.
You may need to hit the button on the status bar to force the app to re-orient to landscape.

### CODE REVIEW

**Comment by mcdurdin** [comment]
Instead of the Keyman logo as the image, could you put a blank 4-row keyboard in -- one without letters on the keys that is representative of any keyboard, and stretch it vertically?

**Review by mcdurdin — CHANGES_REQUESTED** [review]
I have not yet tested this on a device, but conceptually it's looking good.

Can we move `Enable Keyman` and `Set Keyman as default keyboard` above `Adjust keyboard height` and `Spacebar caption`? We are starting to get option proliferation!

**Comment by MakaraSok** [comment]
Please notify me once this PR is ready for testing. It seems like there are still things to be done there per Marc's comment.

File: android/KMAPro/kMAPro/src/main/java/com/tavultesoft/kmapro/AdjustKeyboardHeightActivity.java [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+
+  /**
+   * Refresh the layout for the sample keyboard
+   * @param context
+   * @param updateCurrentHeight - boolean if true, updates currentHeight
+   */
+  private void refreshSampleKeyboard(Context context, boolean updateCurrentHeight) {
+    if (updateCurrentHeight) {
+      currentHeight = KMManager.getKeyboardHeight(context);
+    }
```
[the remark this answers is not in the document]
Comment (darcywong00) [reply android/KMAPro/kMAPro/src/main/java/com/tavultesoft/kmapro/AdjustKeyboardHeightActivity.java:123]: updated settings

File: android/KMAPro/kMAPro/src/main/java/com/tavultesoft/kmapro/AdjustKeyboardHeightActivity.java [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    sampleKeyboard.setOnTouchListener(new View.OnTouchListener() {
+      @Override
+      public boolean onTouch(View view, MotionEvent event) {
+        int y = (int)event.getY();
+
+        switch(event.getAction()) {
+          case MotionEvent.ACTION_MOVE:
+            // Update currentHeight as the user drags (moves)
+            // Increasing the keyboard height is a negative y
+            currentHeight -= y;
```
[the remark this answers is not in the document]
Comment (darcywong00) [reply android/KMAPro/kMAPro/src/main/java/com/tavultesoft/kmapro/AdjustKeyboardHeightActivity.java:97]: given a `defaultHeight` from the layout.xml files I'll gate `currentHeight` so
`defaultHeight / 2` <= `currentHeight` <= `2 * defaultHeight`

**Comment by darcywong00** [comment]
> Instead of the Keyman logo as the image, could you put a blank 4-row keyboard in -- one without letters on the keys that is representative of any keyboard, and stretch it vertically?

Here's an updated screenshot with the stretched blank OSK. I'll need to generate an image to use on tablets...

adjusted blank osk

File: android/KMAPro/kMAPro/src/main/java/com/tavultesoft/kmapro/AdjustKeyboardHeightActivity.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (mcdurdin) [reply android/KMAPro/kMAPro/src/main/java/com/tavultesoft/kmapro/AdjustKeyboardHeightActivity.java]: This LGTM

**Comment by mcdurdin** [comment]
Stretched keyboard looks good

**Review by mcdurdin — APPROVED** [review]
LGTM. Ready for testing!

**Comment by MakaraSok** [comment]
## User Testing Results
| Notes |
| --- |
| Checked, **PASSED** -- works as expected/indicated; otherwise, **FAILED**. |
| Any noticeable behavior would be included as a comment underneath the case in question |
| This test doesn't touch on whether or not each key works as expected, but solely focus on the height adjustments. |

### Setup
* - [x] Load the test build
* - [x] From the "Get Started" menu, enable Keyman as a system-wide keyboard and set Keyman as a default system keyboard.
* Note on the menu for adjusting keyboard height, given a default height (from the layout dimens.xml files, the range of values the keyboard height can be adjusted is:
`default height / 2` <= `adjusted height` <= `2 * default height`

### Adjust Keyboard Height (portrait orientation)
* - [x] Have the device in portrait orientation
* - [x] Go to "Settings" --> "Adjust keyboard height"
* - [x] Note the current keyboard height (default value)
* - [x] On the menu, drag the keyboard to either increase or decrease the keyboard height, and then release your finger
* - [x] Verify the red keyboard height adjusts accordingly
* - [x] Hit the back button until you're back in the Keyman app text area
* - [x] Verify the in-app keyboard height is adjusted per the menu
> _It would be worth noting that the height would exclude the suggestion banner. The banner would be shown right above the resized layout height._
* - [x] On the device, open a separate app (like Chrome)
* - [x] Select an input field and display the Keyman keyboard
* - [x] Verify the system keyboard height is also adjusted per the menu
* - [x] In the Keyman app, go back to the "Adjust keyboard height" menu
* - [x] Press the "RESET TO DEFAULTS" button
* - [x] Verify the red keyboard height adjusts back to the default height
* - [x] Hit the back button until you're back in the Keyman app text area
* - [x] Verify the in-app keyboard height is adjusted per the menu

---

### Adjust Keyboard Height (landscape orientation)
* - [x] Have the device in landscape orientation
* - [x] Go to "Settings" --> "Adjust keyboard height"
* - [x] Note the current keyboard height (default value)
* - [x] On the menu, drag the keyboard to either increase or decrease the keyboard height, and then release your finger
* - [x] Verify the red keyboard height adjusts accordingly
* - [x] Hit the back button until you're back in the Keyman app text area
* - [x] Verify the in-app keyboard height is adjusted per the menu
* - [x] On the device, open a separate app (like Chrome)
* - [x] Select an input field and display the Keyman keyboard
* - [x] Verify the system keyboard height is also adjusted per the menu
* - [x] In the Keyman app, go back to the "Adjust keyboard height" menu
* - [x] Press the "RESET TO DEFAULTS" button
* - [ ] Verify the red keyboard height adjusts back to the default height
* - [ ] Hit the back button until you're back in the Keyman app text area
* - [ ] Verify the in-app keyboard height is adjusted per the menu

> _A general note on portrait/landscape mode, the setting menu shows a stretched keyboard layout where the functions keys look a little ugly. _
>
> _portrait:_ 
> image
>
> _landscape:_ 
> image

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 15.0.99-alpha](https://keyman.com/downloads/releases/alpha/15.0.99)

**Pull request merged**